persist/fs/commitlog/reader.go

```diff
@@ -182,36 +339,44 @@ func (r *reader) readChunk() ([]byte, error) {
 	if _, err := r.chunkReader.Read(buffer); err != nil {
 		return nil, err
 	}
-	return buffer, nil
+
+	// Copy message into pooled byte slice. Reading happens synchronously so its
+	// safe to reuse the same buffer repeatedly, but decoding happens in parallel
+	// so we need to make sure that the byte slice we return won't be mutated
+	// until the caller is done with it.
+	pooled := r.bytesPool.Get(int(size))
+	pooled.IncRef()
```

Comment:
Instead of allowing the caller to DecRef, it's better practice to:
```go
b.IncRef()
b.AppendAll(...)
b.DecRef()
```

And let the consumer then Inc and Dec during use.

Reply:
Yeah I think this is falling out of me not understanding this paradigm properly

Reply:
Maybe add a e2e unittest (mini-integration test) with the leak detection turned on to catch issues like this

Reply:
Do you mean the race detector? or is there a memory leak detector I can turn on or something?

Reply:
Here I mean a memory leak detector. I think there's a way to turn it on for `checked.Bytes`. @robskillington do you remember how?